File: src/main/java/com/procurement/notice/model/tender/record/Record.java
package com.procurement.notice.model.tender.record;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.annotation.JsonPropertyOrder;
import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
import com.fasterxml.jackson.databind.annotation.JsonSerialize;
import com.procurement.notice.databinding.LocalDateTimeDeserializer;
import com.procurement.notice.databinding.LocalDateTimeSerializer;
import com.procurement.notice.model.ocds.*;
import java.time.LocalDateTime;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import lombok.Getter;
import lombok.Setter;

@Getter
@Setter
@JsonInclude(JsonInclude.Include.NON_NULL)
@JsonPropertyOrder({
        "ocid",
        "id",
        "date",
        "tag",
        "initiationType",
        "language",
        "hasPreviousNotice",
        "purposeOfNotice",
        "tender",
        "parties",
        "bids",
        "awards",
        "relatedProcesses"
})
public class Record {
    @JsonProperty("ocid")
    private String ocid;

    @JsonProperty("id")
    private String id;

    @JsonSerialize(using = LocalDateTimeSerializer.class)
    @JsonDeserialize(using = LocalDateTimeDeserializer.class)
    @JsonProperty("date")
    private LocalDateTime date;

    @JsonProperty("tag")
    private List<Tag> tag;

    @JsonProperty("initiationType")
    private InitiationType initiationType;

    @JsonProperty("language")
    private String language;

    @JsonProperty("hasPreviousNotice")
    private final Boolean hasPreviousNotice;

    @JsonProperty("purposeOfNotice")
    private final PurposeOfNotice purposeOfNotice;

    @JsonProperty("tender")
    private final RecordTender tender;

    @JsonProperty("parties")
    private final Set<Organization> parties;

    @JsonProperty("bids")
    private Bids bids;

    @JsonProperty("awards")
    private Set<Award> awards;

    @JsonProperty("relatedProcesses")
    private Set<RelatedProcess> relatedProcesses;

    @JsonCreator
    public Record(@JsonProperty("ocid") final String ocid,
                  @JsonProperty("id") final String id,
                  @JsonProperty("date") final LocalDateTime date,
                  @JsonProperty("tag") final List<Tag> tag,
                  @JsonProperty("initiationType") final InitiationType initiationType,
                  @JsonProperty("language") final String language,
                  @JsonProperty("parties") final HashSet<Organization> parties,
                  @JsonProperty("tender") final RecordTender tender,
                  @JsonProperty("awards") final HashSet<Award> awards,
                  @JsonProperty("bids") final Bids bids,
                  @JsonProperty("hasPreviousNotice") final Boolean hasPreviousNotice,
                  @JsonProperty("purposeOfNotice") final PurposeOfNotice purposeOfNotice,
                  @JsonProperty("relatedProcesses") final HashSet<RelatedProcess> relatedProcesses) {
        this.ocid = ocid;
        this.id = id;
        this.date = date;
        this.tag = tag;
        this.initiationType = initiationType;
        this.parties = parties;
        this.tender = tender;
        this.awards = awards;
        this.language = language == null ? "en" : language;
        this.bids = bids;
        this.hasPreviousNotice = hasPreviousNotice == null ? false : hasPreviousNotice;
        this.purposeOfNotice = purposeOfNotice == null ?
                new PurposeOfNotice(null, false, null)
                : purposeOfNotice;
        this.relatedProcesses = relatedProcesses;
    }
}

File: src/main/kotlin/com/procurement/notice/model/tender/record/Record.kt
package com.procurement.notice.model.tender.record

import com.fasterxml.jackson.annotation.JsonCreator
import com.fasterxml.jackson.annotation.JsonInclude
import com.fasterxml.jackson.annotation.JsonProperty
import com.procurement.notice.model.ocds.Award
import com.procurement.notice.model.ocds.Bids
import com.procurement.notice.model.ocds.Contract
import com.procurement.notice.model.ocds.InitiationType
import com.procurement.notice.model.ocds.Organization
import com.procurement.notice.model.ocds.PurposeOfNotice
import com.procurement.notice.model.ocds.RelatedProcess
import com.procurement.notice.model.ocds.Tag
import java.time.LocalDateTime
import java.util.*

data class Record @JsonCreator constructor(

    @field:JsonInclude(JsonInclude.Include.NON_NULL)
    var ocid: String?,

    @field:JsonInclude(JsonInclude.Include.NON_NULL)
    var id: String?,

    @field:JsonInclude(JsonInclude.Include.NON_NULL)
    var date: LocalDateTime?,

    @field:JsonInclude(JsonInclude.Include.NON_EMPTY)
    var tag: List<Tag>?,

    @field:JsonInclude(JsonInclude.Include.NON_NULL)
    var initiationType: InitiationType?,

    @field:JsonInclude(JsonInclude.Include.NON_EMPTY)
    var parties: HashSet<Organization>?,

    var tender: RecordTender,

    @field:JsonInclude(JsonInclude.Include.NON_EMPTY)
    var awards: HashSet<Award>?,

    @field:JsonInclude(JsonInclude.Include.NON_NULL)
    var bids: Bids?,

    @field:JsonInclude(JsonInclude.Include.NON_EMPTY)
    var contracts: HashSet<Contract>?,

    @get:JsonProperty("hasPreviousNotice")
    @get:JsonInclude(JsonInclude.Include.NON_NULL)
    var hasPreviousNotice: Boolean?,

    @field:JsonInclude(JsonInclude.Include.NON_NULL)
    var purposeOfNotice: PurposeOfNotice?,

    @field:JsonInclude(JsonInclude.Include.NON_EMPTY)
    var relatedProcesses: HashSet<RelatedProcess>?
)
